### PR TITLE
Fix race condition in Jitify

### PR DESCRIPTION
**Pull request by kmaehashi**
This fixes a race condition in jitify that `cupy_headers` (static variable) being modified under `with nogil:`. Reported in https://github.com/cupy/cupy/issues/6793#issuecomment-1308791091 and https://github.com/cupy/cupy/issues/6793#issuecomment-1360368533.

Reproducer

```
$ export CUPY_CACHE_IN_MEMORY=1
$ ./test.py 10
```

`test.py`:

```py
import sys
import threading
import cupy


def run():
    kern = cupy.RawKernel(code='''
#include <cupy/cuda_workaround.h>
#include <cupy/cub/cub/block/block_reduce.cuh>
#include <cupy/cub/cub/block/block_load.cuh>

extern "C" __global__ void mykernel() {}''', name='mykernel', jitify=True)
    kern.compile()

threads = []
for i in range(int(sys.argv[1])):
    threads.append(threading.Thread(target=run))

for t in threads:
    t.start()

for t in threads:
    t.join()
```

### CODE REVIEW

**Review by emcastillo — APPROVED** [review]

File: cupy/cuda/compiler.py [hunk 1]
```diff
@@ -234,11 +234,7 @@ def _jitify_prep(source, options, cu_path):
     global _jitify_header_source_map_populated
     if not _jitify_header_source_map_populated:
         from cupy._core import core
-        _jitify_header_source_map = core._get_header_source_map()
-        _jitify_header_source_map_populated = True
-    else:
-        # this is already cached at the C++ level, so don't pass in anything
-        _jitify_header_source_map = None
+        jitify._add_sources(core._get_header_source_map())
```
Comment (leofang) [inline cupy/cuda/compiler.py:237]: It seems `_jitify_header_source_map_populated` will never be set to `True` with the current change?

**Review by leofang — COMMENTED** [review]
Thanks, @kmaehashi, that's a good catch. I missed it 😓 

Question: If this is true
> This fixes a race condition in jitify that `cupy_headers` (static variable) being modified under `with nogil:`.

shouldn't we just run call the Jitify wrappers `with gil` and be done with it?

**Comment by kmaehashi** [comment]
> shouldn't we just run call the Jitify wrappers with gil and be done with it?

This also works but `load_program` is one of the heaviest parts so I think it's better to make it run in parallel.

File: cupy/cuda/compiler.py (hunk 1 above)
Comment (kmaehashi) [reply cupy/cuda/compiler.py]: Fixed! :+1:

**Comment by leofang** [comment]
I see. So IIUC with this PR we have a few guards against a race:
1. `_add_sources()` in `compiler.py` is protected by GIL
2. before calling `load_program()`, each thread has its local copy of the global, static dict
3. after calling `load_program()`, the global dict is updated but protected by GIL

**Review by leofang — APPROVED** [review]

**Comment by emcastillo** [comment]
```
ERROR cupyx_tests\scipy_tests\sparse_tests\test_csr.py::TestCsrMatrix::test_reshape_0[_param_3_{dtype=complex128}] - cupy_backends.cuda.api.runtime.CUDARuntimeError: cudaErrorIllegalAddress: an illegal memory access was encountered
```
CUDA 10.2 & windows

**Comment by kmaehashi** [comment]
Seems that the problem is in the master branch. I've asked @takagi to investigate.

**Pull request merged**